twake/frontend/src/app/components/ChatUploads/ChatUploadService.ts

```diff
@@ -0,0 +1,162 @@
+  }
+
+  public async upload(fileList: File[]): Promise<void> {
+    const { companyId } = RouterServices.getStateFromRoute();
+
+    if (!fileList) return;
+
+    if (!this.pendingFiles.some(f => isPendingFileStatusPending(f.state.status))) {
+      //New upload task when all previous task is finished
+      this.currentTaskId = uuid();
+    }
+
+    fileList.forEach(async file => {
+      if (!file) return;
+
+      const pendingFile: PendingFileType = {
+        state: {
+          id: uuid(),
+          file: null,
+          status: 'pending',
+          progress: 0,
+        },
+        uploadTaskId: this.currentTaskId,
+        originalFile: file,
+        resumable: null,
+      };
+
+      this.pendingFiles.push(pendingFile);
+
+      this.onChange();
+
+      // First we create the file object
+      const uploadFileRoute = `${this.prefixUrl}/companies/${companyId}/files?filename=${file.name}&type=${file.type}&total_size=${file.size}`;
+      const resource = (await Api.post<undefined, ResponseFileType>(uploadFileRoute, undefined))
+        ?.resource;
+      if (!resource) {
+        throw new Error('A server error occured');
+      }
+
+      pendingFile.state.file = resource;
+      this.onChange();
+
+      // Then we overwrite the file object with resumable
+      pendingFile.resumable = this.getResumableInstance({
+        target: Api.route(
+          `${this.prefixUrl}/companies/${companyId}/files/${pendingFile.state.file.id}`,
+        ),
+        query: {
+          thumbnail_sync: 1,
+        },
+        headers: {
+          Authorization: JWTStorage.getAutorizationHeader(),
+        },
+      });
+
+      pendingFile.resumable.addFile(file);
+
+      pendingFile.resumable.on('fileAdded', (f: any, message: any) =>
+        pendingFile.resumable.upload(),
+      );
+
+      pendingFile.resumable.on('fileProgress', (f: any, ratio: number) => {
+        pendingFile.state.file = f;
+        pendingFile.state.progress = f.progress();
+        this.onChange();
+      });
+
+      pendingFile.resumable.on('fileSuccess', (f: any, message: string) => {
+        pendingFile.state.file = JSON.parse(message).resource;
+        pendingFile.state.status = 'success';
+        this.onChange();
+      });
+
+      pendingFile.resumable.on('fileError', (f: any, message: any) => {
+        pendingFile.state.status = 'error';
+        pendingFile.resumable.cancel();
+        this.onChange();
+      });
+    });
+  }
+
+  public getPendingFile(id: string): PendingFileType {
+    return this.pendingFiles.filter(f => f.state.id === id)[0];
+  }
+
+  public cancel(id: string) {
+    const fileToCancel = this.pendingFiles.filter(f => f.state.id === id)[0];
+
+    fileToCancel.resumable.cancel();
+    fileToCancel.state.status = 'error';
+    this.onChange();
+
+    setTimeout(() => {
+      this.pendingFiles = this.pendingFiles.filter(f => f.state.id !== id);
+      this.onChange();
+    }, 1000);
```

Comment:
Really? What is the goal of this?
If you really need a timeout, you should add a parameter to the function like

```js
public cancel(id: string, timeout = 1000) {

}
```

Reply:
Updated